components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/TenancyLoader.java

```diff
@@ -0,0 +1,153 @@
+import org.wso2.carbon.apimgt.impl.dto.GatewayArtifactSynchronizerProperties;
+import org.wso2.carbon.apimgt.impl.dto.LoadingTenants;
+import org.wso2.carbon.apimgt.impl.utils.APIUtil;
+import org.wso2.carbon.core.ServerStartupObserver;
+import org.wso2.carbon.stratos.common.exception.TenantMgtException;
+import org.wso2.carbon.tenant.mgt.services.TenantMgtService;
+import org.wso2.carbon.user.api.UserStoreException;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This class is responsible for loading tenant information from the Event Hub to Gateway.
+ */
+public class TenancyLoader implements ServerStartupObserver {
+    private static final Log log = LogFactory.getLog(TenancyLoader.class);
+    private final EventHubConfigurationDto eventHubConfigurationDto = ServiceReferenceHolder.getInstance().getAPIManagerConfiguration().getEventHubConfigurationDto();
+    private final String baseURL = eventHubConfigurationDto.getServiceUrl() + APIConstants.INTERNAL_WEB_APP_EP;
+
+    public void retrieveAndLoadAllTenants() throws APIManagementException {
+        String base64EncodedTenants;
+        try {
+            base64EncodedTenants = Base64.encodeBase64URLSafeString(
+                    retrieveTenantFilter().getBytes(APIConstants.DigestAuthConstants.CHARSET));
+            String endpoint = baseURL + APIConstants.TENANT_INFO_ENDPOINT + "?filter=" + base64EncodedTenants;
+            try (CloseableHttpResponse closeableHttpResponse = invokeService(endpoint)) {
+                if (closeableHttpResponse.getStatusLine().getStatusCode() == 200) {
+                    try (InputStream inputStream = closeableHttpResponse.getEntity().getContent()) {
+                        String response = IOUtils.toString(inputStream, APIConstants.DigestAuthConstants.CHARSET);
+                        TenantInfoList tenantInfoList = new Gson().fromJson(response, TenantInfoList.class);
+                        TenantMgtService tenantMgtService = ServiceReferenceHolder.getInstance().getTenantMgtService();
+                        for (TenantInfo tenantInfo : tenantInfoList.getTenants()) {
+                            try {
+                                if (tenantMgtService.isDomainAvailable(tenantInfo.getDomain())) {
+                                    TenantUtils.addTenant(tenantInfo);
+                                } else {
+                                    TenantUtils.updateTenant(tenantInfo);
+                                }
+                            } catch (TenantMgtException | UserStoreException e) {
+                                log.error("Error occurred while registering tenant " + tenantInfo.getDomain(), e);
+                            }
+                        }
+                    }
+                } else {
+                    throw new APIManagementException("Failed to retrieve tenant information");
```

Comment:
Shall we add error log containing the failure information. since we only check for status code 200 for the if condition, we could at least log the status code for failure case to get an idea about the fault